File: python/mirage/commands/cli/builtin/gh/release.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.commands.cli.builtin.gh.accessor import (body_value, camel,
                                                     csv_values, list_limit,
                                                     repo_for, text_out,
                                                     typed_out)
from mirage.commands.cli.types import CLIInvocation
from mirage.commands.spec.flag_view import FlagView
from mirage.core.github.config import GhConfig
from mirage.core.github.release import (create_release, get_latest_release,
                                        get_release, list_releases)
from mirage.io.types import ByteSource, IOResult
from mirage.types import JsonValue

RELEASE_FIELDS = ("body", "createdAt", "isDraft", "isLatest", "isPrerelease",
                  "name", "publishedAt", "tagName", "targetCommitish", "url")


def _release(value: JsonValue) -> dict[str, Any]:
    row = camel(value)
    result = row if isinstance(row, dict) else {}
    if "draft" in result:
        result["isDraft"] = result.pop("draft")
    if "prerelease" in result:
        result["isPrerelease"] = result.pop("prerelease")
    result.setdefault("isLatest", False)
    return result


def _list_line(row: dict[str, Any]) -> str:
    kind = ("Draft" if row.get("isDraft") else "Pre-release" if row.get(
        "isPrerelease") else "Latest" if row.get("isLatest") else "")
    return (f'{row.get("name", "")}\t{row.get("tagName", "")}\t{kind}\t'
            f'{row.get("publishedAt", "")}\n')


def _list_text(rows: list[dict[str, Any]]) -> str:
    return "".join(_list_line(row) for row in rows)


async def list_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    rows = [
        _release(value) for value in await list_releases(
            inv.config, repo_for(inv, fl), list_limit(fl, 30))
    ]
    latest = next((row for row in rows
                   if not row.get("isDraft") and not row.get("isPrerelease")),
                  None)
    if latest is not None:
        latest["isLatest"] = True
    return await typed_out(rows, fl, _list_text(rows), RELEASE_FIELDS)


async def view_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    tag = inv.texts[0] if inv.texts else ""
    if not tag:
        raise ValueError("a release tag is required")
    ref = repo_for(inv, fl)
    row = _release(await get_release(inv.config, ref, tag))
    if "isLatest" in csv_values([fl.as_str("json") or ""]):
        latest = await get_latest_release(inv.config, ref)
        latest_row = _release(latest) if latest is not None else {}
        row["isLatest"] = latest_row.get("tagName") == row.get("tagName")
    human = (f'title:\t{row.get("name", "")}\n'
             f'tag:\t{row.get("tagName", "")}\n'
             f'draft:\t{str(bool(row.get("isDraft"))).lower()}\n'
             f'prerelease:\t{str(bool(row.get("isPrerelease"))).lower()}\n--\n'
             f'{row.get("body", "")}\n')
    return await typed_out(row, fl, human, RELEASE_FIELDS)


async def create_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    tag = inv.texts[0] if inv.texts else ""
    if not tag:
        raise ValueError("a release tag is required in noninteractive mode")
    notes = await body_value(inv, fl, value="notes", file="notes_file")
    body: dict[str, JsonValue] = {
        "tag_name": tag,
        "name": fl.as_str("title") or tag,
        "body": notes or "",
        "draft": fl.as_bool("draft"),
        "prerelease": fl.as_bool("prerelease"),
        "generate_release_notes": fl.as_bool("generate_notes"),
    }
    if fl.as_str("target") is not None:
        body["target_commitish"] = fl.as_str("target") or ""
    created = _release(await create_release(inv.config, repo_for(inv, fl),
                                            body))
    return text_out(f'{created.get("url", "")}\n')

File: typescript/packages/core/src/commands/cli/builtin/gh/release.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { FlagView } from '../../../spec/flag_view.ts'
import type { CommandFnResult } from '../../../config.ts'
import type { CLIInvocation } from '../../types.ts'
import {
  createRelease,
  getLatestRelease,
  getRelease,
  listReleases,
} from '../../../../core/github/release.ts'
import {
  bodyValue,
  camel,
  csvValues,
  ghTransport,
  repoFor,
  textOut,
  textValue,
  typedOut,
} from './accessor.ts'

const RELEASE_FIELDS = [
  'body',
  'createdAt',
  'isDraft',
  'isLatest',
  'isPrerelease',
  'name',
  'publishedAt',
  'tagName',
  'targetCommitish',
  'url',
] as const

function release(value: unknown): Record<string, unknown> {
  const row = camel(value)
  const result = row !== null && typeof row === 'object' ? (row as Record<string, unknown>) : {}
  if ('draft' in result) {
    result.isDraft = result.draft
    delete result.draft
  }
  if ('prerelease' in result) {
    result.isPrerelease = result.prerelease
    delete result.prerelease
  }
  result.isLatest ??= false
  return result
}

function listText(rows: Record<string, unknown>[]): string {
  return rows
    .map((row) => {
      const kind = row.isDraft
        ? 'Draft'
        : row.isPrerelease
          ? 'Pre-release'
          : row.isLatest
            ? 'Latest'
            : ''
      return `${textValue(row.name)}\t${textValue(row.tagName)}\t${kind}\t${textValue(row.publishedAt)}\n`
    })
    .join('')
}

export async function listCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const rows = (
    await listReleases(ghTransport(inv.config), repoFor(inv, fl), fl.asInt('limit') ?? 30)
  ).map(release)
  const latest = rows.find((row) => !row.isDraft && !row.isPrerelease)
  if (latest !== undefined) latest.isLatest = true
  return typedOut(rows, fl, listText(rows), RELEASE_FIELDS)
}

export async function viewCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const tag = inv.texts[0] ?? ''
  if (tag === '') throw new Error('a release tag is required')
  const transport = ghTransport(inv.config)
  const ref = repoFor(inv, fl)
  const row = release(await getRelease(transport, ref, tag))
  if (csvValues([fl.asStr('json') ?? '']).includes('isLatest')) {
    const latest = await getLatestRelease(transport, ref)
    const latestRow = latest === null ? {} : release(latest)
    row.isLatest = latestRow.tagName === row.tagName
  }
  const human = `title:\t${textValue(row.name)}\ntag:\t${textValue(row.tagName)}\ndraft:\t${String(Boolean(row.isDraft))}\nprerelease:\t${String(Boolean(row.isPrerelease))}\n--\n${textValue(row.body)}\n`
  return typedOut(row, fl, human, RELEASE_FIELDS)
}

export async function createCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const tag = inv.texts[0] ?? ''
  if (tag === '') throw new Error('a release tag is required in noninteractive mode')
  const body: Record<string, unknown> = {
    tag_name: tag,
    name: fl.asStr('title') ?? tag,
    body: (await bodyValue(inv, fl, { value: 'notes', file: 'notes_file' })) ?? '',
    draft: fl.asBool('draft'),
    prerelease: fl.asBool('prerelease'),
    generate_release_notes: fl.asBool('generate_notes'),
  }
  const target = fl.asStr('target')
  if (target !== undefined) body.target_commitish = target
  const created = release(await createRelease(ghTransport(inv.config), repoFor(inv, fl), body))
  return textOut(`${textValue(created.url)}\n`)
}
